### PR TITLE
Fix #2415: add live build logs in status dashboard

### LINKED ISSUE #2415: Add live build logs in status dashboard
In #2395 we are adding an expanded status service dashboard. One thing I want to include in it is the ability to see live build logs from the [autodeployment server](https://github.com/Seneca-CDOT/telescope/tree/master/tools/autodeployment), which are exposed at `/log`, for example:

- staging is https://dev.telescope.cdot.systems:4000/log
- production is https://telescope.cdot.systems:4000/log

If there's no current build, it returns a 404, otherwise it streams the build log live.  I'd like to use [xterm.js](https://xtermjs.org/) to show a terminal with live output in the dashboard, similar to what happens when you look at build output for GitHub Actions or other CI builders.

I've started doing this in the past in React, but never finished the code, see my [autodeploy-ui branch](https://github.com/humphd/telescope/tree/autodeploy-ui), and [these files](https://github.com/humphd/telescope/tree/autodeploy-ui/tools/autodeployment/src/web) in particular.  For the new dashboard, this needs to get rewritten to be straight HTML/CSS/JS, but the logic for how I use xterm and the live build log stream is something you could borrow:

https://github.com/humphd/telescope/blob/autodeploy-ui/tools/autodeployment/src/web/components/BuildLog.js

### PR DESCRIPTION
**Pull request by Andrewnt219**
## Issue This PR Addresses

Closes #2415

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
On /build.html, periodically fetches build status and shows build logs (if any) using [xterm](https://xtermjs.org/).   

image

image

## Test plan

To test locally, change to `<base href="/" />` inside build.html and run start script. The site is on `localhost:1111/pages/build.html`.

Xterm UI should show up. In the dev console, `/status` is requested periodically. 
## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

<a href="https://gitpod.io/#https://github.com/Seneca-CDOT/telescope/pull/2455"></a>

### CODE REVIEW

File: src/api/status/public/build.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    <!-- Material Icons -->
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons+Round" rel="stylesheet" />
+    <!-- CSS Files -->
+    <link id="pagestyle" href="assets/css/material-dashboard.css?v=3.0.0" rel="stylesheet" />
+  </head>
+
+  <body class="g-sidenav-show bg-gray-200">
+    <div id="terminal"></div>
+
+    <script type="module" src="buildLog/index.js" defer></script>
```
Comment (humphd) [inline src/api/status/public/build.html:35]: Probably worth moving this below the core files under a new comment that indicates it's part of Telescope stuff:

```html
<!--   Core JS Files   -->
...

<!--   Telescope JS Files   -->
...
```

File: src/api/status/public/buildLog/api.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+export async function checkBuildStatus() {
+  try {
+    const res = await fetch('/v1/status/status');
```
Comment (humphd) [inline src/api/status/public/buildLog/api.js:3]: This is going to need to point to the autodeployment server.

We're probably going to need to do some URL trickery to figure out which server (local vs. staging vs. production) to use, where you take the current URL and add `:4000/status` to the end.

File: src/api/status/public/buildLog/api.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    };
+  } catch (err) {
+    console.error(err);
+    return { building: false };
+  }
+}
+
+export async function getBuildLog() {
+  try {
+    const res = await fetch('/v1/status/log');
```
Comment (humphd) [inline src/api/status/public/buildLog/api.js:25]: Same thing here, needs to be `<current-url>:4000/log`

File: src/api/status/public/buildLog/terminal.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  brightWhite: '#F8F8F2',
+  brightYellow: '#FEED6C',
+  cyan: '#56C2D6',
+  green: '#82B414 ',
+  magenta: '#8C54FE',
+  red: '#F92672',
+  white: '#CCCCC6',
+  yellow: '#FD971F',
+};
+const terminal = new Terminal({
```
Comment (humphd) [inline src/api/status/public/buildLog/terminal.js:22]: You're missing `Terminal` and the xterm addon stuff?

**Review by humphd — CHANGES_REQUESTED** [review]
We need to fix #2454 for the dev setup.

Also, let's get a link from the main dashboard `index.html` to `build.html`.

File: src/api/status/public/buildLog/terminal.js (hunk 4 above)
Comment (Andrewnt219) [reply src/api/status/public/buildLog/terminal.js:22]: Are they not these ones on L29-34?

```js
const fitAddon = new FitAddon.FitAddon();
terminal.loadAddon(fitAddon);
terminal.open(document.getElementById('terminal'));
terminal.loadAddon(new WebLinksAddon.WebLinksAddon());
terminal.loadAddon(new WebglAddon.WebglAddon());
fitAddon.fit();
```

File: src/api/status/public/buildLog/api.js (hunk 2 above)
Comment (Andrewnt219) [reply src/api/status/public/buildLog/api.js:3]: So I guess the only way to test it is pushing to dev?

File: src/api/status/public/js/build-log/api.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+function getCurrentUrl() {
+  // http://localhost:1111/pages/build.html -> http://localhost:4000/status
+  return `${location.protocol}//${location.hostname}:4000/status`;
```
Comment (humphd) [inline src/api/status/public/js/build-log/api.js:3]: You can use `//` and not bother with the protocol, it will use the same protocol as the current URL.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/status/public/pages/build.html [hunk 6]
```diff
@@ -0,0 +1,45 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <base href="/v1/status/" />
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
+    <link rel="apple-touch-icon" sizes="76x76" href="assets/img/apple-icon.png" />
+    <link rel="icon" type="image/png" href="assets/img/favicon.png" />
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/xterm@4.15.0/css/xterm.css" />
+    <script src="https://cdn.jsdelivr.net/npm/xterm@4.15.0/lib/xterm.min.js"></script>
```
[the remark this answers is not in the document]
Comment (Andrewnt219) [reply src/api/status/public/pages/build.html:10]: I think it's easy to look up packages on jsdelivr. Their syntax for specifying package versions is the same, but unpkg doesn't have a search page.

```bash
cdn.jsdelivr.net/npm/:package@:version/:file

unpkg.com/:package@:version/:file
```

image

File: src/api/status/public/js/build-log/api.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+function getCurrentUrl() {
```
Comment (humphd) [inline src/api/status/public/js/build-log/api.js:1]: You need this to work for `/status` and `/log` so why don't you do:

```js
const autodeploymentUrl = path => `//${location.hostname}:4000/${path}`;
```

**Review by humphd — CHANGES_REQUESTED** [review]
Looking close to me

File: src/api/status/public/pages/build.html [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    <script src="https://kit.fontawesome.com/42d5adcbca.js" crossorigin="anonymous"></script>
+    <!-- Material Icons -->
+    <link href="https://fonts.googleapis.com/icon?family=Material+Icons+Round" rel="stylesheet" />
+    <!-- CSS Files -->
+    <link id="pagestyle" href="assets/css/material-dashboard.css?v=3.0.0" rel="stylesheet" />
+  </head>
+
+  <body class="g-sidenav-show bg-gray-200">
+    <h1>Build log</h1>
+    <div id="terminal"></div>
```
Comment (humphd) [inline src/api/status/public/pages/build.html:30]: Question: do we need any CSS on this element to have it fill the available space, or does the fit addon do that for us (I can't remember)?

**Review by humphd — APPROVED** [review]

**Comment by humphd** [comment]
@Andrewnt219 can you squash/rebase this for me?  I think it's basically ready.  I just want one more review on it.

cc @manekenpix, in case you want to read it before it goes in.

File: src/api/status/public/pages/build.html (hunk 8 above)
Comment (Andrewnt219) [reply src/api/status/public/pages/build.html:30]: I just copied it from the index page. So we don't really need it.

**Review by humphd — APPROVED** [review]
Can't wait for this to finally happen!  I've been wanting to do it for over a year, and never had time to finish it.  Thanks for taking it on.

**Comment by Andrewnt219** [comment]
This's just the first step. Let's deploy it to see if it's everything's working and I'll continue to iterate over this feature until it's finished 👍.

**Review by manekenpix — APPROVED** [review]

**Pull request merged**